docs/pages/pmd/userdocs/extending/writing_java_rules.md

```diff
@@ -0,0 +1,164 @@
+
+    @Override
+    public Object visit(ASTVariableDeclaratorId node, Object data) {
+        // This method is called on each node of type ASTVariableDeclaratorId
+        // in the AST
+
+        if (node.getType() == short.class) {
+            // reports a violation at the position of the node
+            // the "data" parameter is a context object handed to by your rule
+            // the message for the violation is the message defined in the rule declaration XML element
+            addViolation(data, node);
+        }
+
+        // this calls back to the default implementation, which recurses further down the subtree
+        return super.visit(node, data);
+    }
+}
+```
+
+The `super.visit(node, data)` call is super common in rule implementations,
+because it makes the traversal continue by visiting all the descendants of the
+current node.
+
+#### Stopping the traversal
+
+Sometimes you have checked all you needed and you're sure that the descendants
+of a node may not contain violations. In that case, you can avoid calling the
+`super` implementation and the traversal will not continue further down. This
+means that your callbacks (`visit` implementations) won't be called on the rest
+of the subtree. The siblings of the current node may be visited
+recursively nevertheless.
+
+#### Economic traversal: the rulechain
+
+If you don't care about the order in which the nodes are traversed (e.g. your
+rule doesn't maintain any state between visits), then you can monumentally
+speed-up your rule by using the **rulechain**.
+
+That mechanism doesn't recurse on all the tree, instead, your rule will only be
+passed the nodes it is interested in. To use the rulechain correctly:
+* Your rule must register those node types by calling {% jdoc core::Rule#addRuleChainVisit(java.lang.Class) %}
+in its constructor.
+* Your visit methods **must not recurse!** In effect, you should call never
+call `super.visit` in the methods.
+
+### Execution across files, thread-safety and statefulness
+
+When starting execution, PMD will instantiate a new instance of your rule.
+If PMD is executed in multiple threads, then each thread is using its own
+instance of the rule. This means, that the rule implementation **does not need to care about
+threading issues**, as PMD makes sure, that a single instance is not used concurrently
+by multiple threads.
+
+However, for performance reasons, the rule instances are used for multiple files.
+This means, that the constructor of the rule is only executed once (per thread)
+and the rule instance is reused. If you rely on a proper initialization of instance
+properties, you can do the initialization e.g. in the visit-method of the {% jdoc jast::ASTCompilationUnit %}
+node - which is visited first and only once per file. However, this
+solution would only work for rules written for the Java language. A language
+independent way is to override the method `apply` of the rule (and call super).
```

Comment:
I think, it's actually better to override the `start(RuleContext)` method, to do any initialization before a file is analyzed.
I'll change that...